doc/languages-frameworks/agda.section.md

```diff
@@ -0,0 +1,85 @@
+---
+title: Agda
+author: Alex Rice (alexarice)
+date: 2020-01-06
+---
+# Agda
+
+## How to use Agda
+
+Agda can be installed with `agda.Agda`. For example:
+```
+$ nix-env -iA agda.Agda
+```
+
+To use agda with libraries, the `withPackages` function can be used. This function either takes:
++ A list of packages,
++ or a function which returns a list of packages when given the `agda` attribute set,
++ or an attribute set containing a list of packages and local home libraries (see below).
+
+For example, suppose we wanted a version of agda which has access to the standard library. This can be obtained with the expressions:
+
+```
+agda.withPackages [ agda.standard-library ]
+```
+
+or
+
+```
+agda.withPackages (p: [ p.standard-library ])
+```
+
+If you want to use a library in your home directory (for instance if it doesn't have a nix derivation or is a development version) then an attribute set can be passed into `agda.withPackages` with the following arguments:
+
+```
+agda.withPackages {
+  pkgs = [ agda.standard-library ];
+  homeLibraries = "path/to/my/package.agda-lib";
```

Comment:
It does do this.

Reply:
I don't think it's a good idea to introduce such an interface. At the very least you should not advertise it on the very first page of the agda manual, before covering how to do the "proper" pure packaging.

Reply:
I seem to remember thinking that this was important at some point, though it seems less so now. @laMudri  @turion do you have any thoughts on this?

Reply:
I think it was included as using master versions of libraries is quite common in Agda

Reply:
Why should they be rebuilt every time. Many people will be developing those?

Reply:
I'm starting to think this isn't needed. An alternative is overriding the source for the package with the a local checkout in home (which has already been built). This has pretty much the same effect

Reply:
I understood this differently. I thought it creates a file in the store somewhere with content the _string_ `"path/to/my/package.agda-lib"`. So you have a store file containing the path of a file in the user's home, but not that impure file itself. Am I wrong here?

Reply:
This is exactly what it does, but it still has the effect that your "pure" agda installation can break if you change stuff in your home directory

Reply:
I still think this is ok. After all it's also common to specify a secret key file that is not in the store, and if you break that file your installation is broken.

Or else, many programs look for user settings in `~/.foo`. NixOS doesn't make any effort to "fix" that by forcing the program to look for a setting exclusively in the store. It's the same here, with the difference that you can specify that path.

Reply:
> I understood this differently. I thought it creates a file in the store somewhere with content the _string_ `"path/to/my/package.agda-lib"`. So you have a store file containing the path of a file in the user's home, but not that impure file itself. Am I wrong here?

I believe, you are not wrong here in the sense that there is not a one single clear definition of what "pure" with respect to software should mean. This has to do with where you put the boundary between pure bits and impure external world. We, generally, do want to build everything in pure environments, with exceptions when we have to make a *compromise* and settle for impure builds due to limited computing resources.

> After all it's also common to specify a secret key file that is not in the store, and if you break that file your installation is broken.

This is a bad example. The reason why secrets don't go to the store is because the store is currently world-readable. Putting them outside of the store is a workaround to a long standing issue https://github.com/NixOS/nix/issues/8.

> Or else, many programs look for user settings in `~/.foo`. NixOS doesn't make any effort to "fix" that by forcing the program to look for a setting exclusively in the store. It's the same here, with the difference that you can specify that path.

And some people see this as an issue. I think, https://github.com/rycee/home-manager is there to address this.